pkiinit/option/generate.go

```diff
@@ -0,0 +1,200 @@
+}
+
+func generatePkis() (exitCode, error) {
+	baseWorkingDir, err := os.Getwd()
+	if err != nil {
+		return exitWithError, err
+	}
+
+	pkiSetupRunPath := filepath.Join(baseWorkingDir, pkiSetupExecutable)
+	pkiSetupVaultJSONPath := filepath.Join(baseWorkingDir, pkiSetupVaultJSON)
+
+	scratchPath := filepath.Join(os.Getenv(envXdgRuntimeDir), pkiInitScratchDir)
+
+	log.Println("pkiSetupRunPath: ", pkiSetupRunPath,
+		"  pkiSetupVaultJSONPath: ", pkiSetupVaultJSONPath,
+		"  scratchPath: ", scratchPath)
+
+	if _, err := exec.LookPath(pkiSetupRunPath); err != nil {
+		return exitWithError, err
+	}
+
+	if _, err := os.Stat(pkiSetupVaultJSONPath); os.IsNotExist(err) {
+		log.Println("Vault JSON file for pkisetup not exists in ", pkiSetupVaultJSONPath)
+		return exitWithError, err
+	}
+
+	// create scratch dir if not exists yet:
+	if err := createDirectoryIfNotExists(scratchPath); err != nil {
+		return exitWithError, err
+	}
+
+	// after done, need to change it back to the original working dir to avoid os.Getwd() error
+	// and delete the scratch dir
+	defer cleanup(baseWorkingDir, scratchPath)
+
+	// run pkisetup binary on the env. of $XDG_RUNTIME_DIR/edgex/pki-init/scratch
+	if err := os.Chdir(scratchPath); err != nil {
+		return exitWithError, err
+	}
+
+	cmd := exec.Command(pkiSetupRunPath, "--config="+pkiSetupVaultJSONPath)
+	result, execErr := cmd.CombinedOutput()
+	if execErr != nil {
+		return exitWithError, execErr
+	}
+
+	log.Printf("result of pkisetup: %s\n", string(result))
+
+	return rearrangePkiByServices(baseWorkingDir, pkiSetupVaultJSONPath)
+}
+
+func rearrangePkiByServices(baseWorkingDir, pkiSetupVaultJSONPath string) (exitCode, error) {
+	reader := cert.NewX509ConfigReader(pkiSetupVaultJSONPath)
+
+	if reader == nil {
+		return exitWithError, fmt.Errorf("Failed to create X509ConfigReader with json path: %s", pkiSetupVaultJSONPath)
+	}
+
+	config, readErr := reader.Read()
+	if readErr != nil {
+		return exitWithError, readErr
+	}
+
+	pkiOutputDir, err := config.GetPkiOutputDir()
+	if err != nil {
+		return exitWithError, err
+	}
+
+	generatedDirPath := filepath.Join(os.Getenv(envXdgRuntimeDir), pkiInitGeneratedDir)
+
+	log.Println("pki-init generate output base dir: ", generatedDirPath)
+
+	// create generated dir if not exists yet:
+	if err := createDirectoryIfNotExists(generatedDirPath); err != nil {
+		return exitWithError, err
+	}
+
+	// CA:
+	caDirPath := filepath.Join(generatedDirPath, caServiceName)
+	err = copyGeneratedForSerive(caDirPath, pkiOutputDir, config)
+
+	// Vault:
+	vaultServicePath := filepath.Join(generatedDirPath, vaultServiceName)
+	err = copyGeneratedForSerive(vaultServicePath, pkiOutputDir, config)
+
+	return normal, nil
+}
+
+func createDirectoryIfNotExists(dirName string) (err error) {
+	if _, err := os.Stat(dirName); os.IsNotExist(err) {
+		err = os.MkdirAll(dirName, os.ModePerm)
+	}
+	return err
+}
+
+func copyGeneratedForSerive(servicePath, pkiOutputDir string, config cert.X509Config) error {
```

Comment:
rename to copyGeneratedForService(...)
service was spelled incorrectly

Reply:
good catch